app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -24,23 +45,79 @@ open class FeedViewModel : ViewModel() {
     }
 
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
         isEmpty.value = true
+        isLoading.value = true
+
+
+        val retrofit = Retrofit.Builder()
+            .baseUrl(BASE_URL)
+            .addConverterFactory(
+                MoshiConverterFactory.create(Moshi.Builder()
+                    .add(KotlinJsonAdapterFactory())
+                    .add(ThumbnailUrlAdapter())
+                    .build()))
+            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
+            .build()
+
+        val service = retrofit.create(FeedApiService::class.java)
+        service.singleFeedItem()
+            .subscribeOn(Schedulers.io())
+//            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe({ feedResponse ->
+                handleResponse(feedResponse)
+            },{ error ->
+                handleNetworkError(error)
+            })
+    }
+
+    private fun convarteToFeedItemsList(items: List<FeedItemDTO>?): List<FeedItem> {
+        val listFeedItem = ArrayList<FeedItem>()
+        for (item in items!!)
+        {
+            val feedItem = FeedItem(item.id, item.thumbnailUrl, item.isPremium)
+            listFeedItem.add(feedItem)
+        }
+        return listFeedItem
     }
+
+
+    private fun insertToDb(items: List<FeedItemDTO>?){
+        val listEntity = ArrayList<FeedItemEntity>()
+        for (item in items!!)
+        {
+            val feedItemEntity = FeedItemEntity(item.id, item.thumbnailUrl, item.isPremium)
+            listEntity.add(feedItemEntity)
+        }
+        feedDatabase.feedItemDao().insertAll(listEntity).subscribe()
+    }
+
+    private fun handleNetworkError(error: Throwable?) {
+        isLoading.postValue(false)
+        Log.d("TAG", "handleNetworkError: ${error?.message.toString()}" )//+ error.toString())
+    }
+
+    private fun handleResponse(feedResponse: GetFeedResponse?) {
+        Log.d("TAG", "handleResponse: ${feedResponse?.toString()}")
+        val feed = feedResponse?.templatesMetadata
+        insertToDb(feed)
+        feedItems.postValue(convarteToFeedItemsList(feed))
+        isLoading.postValue(false)
+        isEmpty.postValue(feedItems.getValue()?.isEmpty())
```

Comment:
This won't work. In line 103 you call:
`feedItems.postValue(convarteToFeedItemsList(feed))`

This is an asynchronous call. It takes a while until `feedItems` will reflect it.
So in line 105 `feedItems.getValue()` will return the previous value (probably `null`).